compiler/grainformat/comment_utils.re

```diff
@@ -0,0 +1,564 @@
+  | [] => []
+  | [cmt, ...remaining_comments] =>
+    let c_loc: Location.t = Locations.get_comment_loc(cmt);
+
+    let (_, cmtsline, cmtschar, _) =
+      Locations.get_raw_pos_info(c_loc.loc_start);
+
+    let (_, cmteline, cmtechar, _) =
+      Locations.get_raw_pos_info(c_loc.loc_end);
+    if (cmtsline > stmt_end_line) {
+      []; // can stop now
+    } else if (cmteline < stmt_start_line) {
+      get_comments_inside_location(location, remaining_comments);
+    } else if
+      // other cases were simple as we are on lines before or after.
+      // Now we need to check when the start line or end line match that we also take
+      // into account the start or end characte
+      (cmtsline > stmt_start_line
+       || cmtsline == stmt_start_line
+       && cmtschar >= stm_start_char) {
+      if (cmteline < stmt_end_line
+          || cmteline == stmt_end_line
+          && cmtechar <= stmt_end_char) {
+        [cmt, ...get_comments_inside_location(location, remaining_comments)];
+      } else {
+        get_comments_inside_location(location, remaining_comments);
+      };
+    } else {
+      get_comments_inside_location(location, remaining_comments);
+    };
+  };
+};
+
+let get_comments_between_locations =
+    (~loc1: Location.t, ~loc2: Location.t, comments: list(Parsetree.comment)) => {
+  let (_, stmt_end_line, stmt_end_char, _) =
+    Locations.get_raw_pos_info(loc1.loc_end);
+  let (_, stmt_start_line, stm_start_char, _) =
+    Locations.get_raw_pos_info(loc2.loc_start);
+
+  // invert the request to look inside the location in the gap
+
+  let start_loc: Lexing.position = {
+    pos_fname: "",
+    pos_lnum: stmt_end_line,
+    pos_bol: 0,
+    pos_cnum: stmt_end_char,
+  };
+  let end_loc: Lexing.position = {
+    pos_fname: "",
+    pos_lnum: stmt_start_line,
+    pos_bol: 0,
+    pos_cnum: stm_start_char,
+  };
+
+  let location: Location.t = {
+    loc_start: start_loc,
+    loc_end: end_loc,
+    loc_ghost: false,
+  };
+  get_comments_inside_location(~location, comments);
+};
+
+let get_comments_enclosed_and_before_location =
+    (~loc1: Location.t, ~loc2: Location.t, comments: list(Parsetree.comment)) => {
+  let (_, loc_start_line, loc_start_char, _) =
+    Locations.get_raw_pos_info(loc1.loc_start);
+  let (_, stmt_start_line, stm_start_char, _) =
+    Locations.get_raw_pos_info(loc2.loc_start);
+
+  // invert the request to look inside the location in the gap
+
+  let start_loc: Lexing.position = {
+    pos_fname: "",
+    pos_lnum: loc_start_line,
+    pos_bol: 0,
+    pos_cnum: loc_start_char,
+  };
+  let end_loc: Lexing.position = {
+    pos_fname: "",
+    pos_lnum: stmt_start_line,
+    pos_bol: 0,
+    pos_cnum: stm_start_char,
+  };
+
+  let location: Location.t = {
+    loc_start: start_loc,
+    loc_end: end_loc,
+    loc_ghost: false,
+  };
+
+  get_comments_inside_location(~location, comments);
+};
+
+let get_comments_from_start_of_enclosing_location =
+    (
+      ~wrapper: Location.t,
```

Comment:
I felt like `wrapper` was a weird name, should this be something like `outer_location` or `location_of_outer` or something?

Reply:
It's the location that encloses or wraps the code and comments we are looking for, but happy to rename